From 1f69f684c3da7bca1607d3645c26275d2da8ccab Mon Sep 17 00:00:00 2001
From: Paul Stewart <pstew@chromium.org>
Date: Thu, 7 Jun 2012 17:50:36 -0700
Subject: [PATCH] CHROMIUMOS: wpa_supplicant: Don't leak scan frequencies

Use a single buffer for storing the list of frequencies to use
for background scan (in the order they should be scanned).
Instead of allocating it once for each scan (then leaking it,
as the previous implementation did), create the buffer at
the start of the background scan method.

BUG=chromium-os:30050
TEST=valgrind

Change-Id: I743caecde13ffd973262d51b336c255af62835af
Reviewed-on: https://gerrit.chromium.org/gerrit/24829
Reviewed-by: Gary Morain <gmorain@chromium.org>
Commit-Ready: Paul Stewart <pstew@chromium.org>
Tested-by: Paul Stewart <pstew@chromium.org>
---
 wpa_supplicant/bgscan_simple.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/wpa_supplicant/bgscan_simple.c b/wpa_supplicant/bgscan_simple.c
index 42c993e..10c1b65 100644
--- a/wpa_supplicant/bgscan_simple.c
+++ b/wpa_supplicant/bgscan_simple.c
@@ -30,6 +30,7 @@ struct bgscan_simple_data {
 	struct os_reltime last_bgscan;
 	int *supp_freqs;
 	int n_supp_freqs;
+	int *scan_freqs;
 	int freq_idx;
 	struct bgscan_signal_monitor_state signal_monitor;
 };
@@ -37,13 +38,12 @@ struct bgscan_simple_data {
 
 static int * bgscan_simple_get_freqs(struct bgscan_simple_data *data)
 {
-	int *freqs;
+	int *freqs = data->scan_freqs;
 	int i, j;
 
 	if (data->supp_freqs == NULL)
 		return NULL;
 
-	freqs = os_malloc((data->n_supp_freqs + 1) * sizeof(int));
 	if (freqs == NULL)
 		return NULL;
 
@@ -218,6 +218,7 @@ static void bgscan_simple_setup_freqs(struct wpa_supplicant *wpa_s,
 		/* TODO(sleffler) priority order freqs */
 		data->supp_freqs = freqs;
 		data->n_supp_freqs = count;
+		data->scan_freqs = os_malloc((count + 1) * sizeof(int));
 
 		log_freqs("Supported", freqs);
 	}
@@ -290,6 +291,7 @@ static void bgscan_simple_deinit(void *priv)
 	if (data->signal_threshold)
 		bgscan_deinit_signal_monitor(&data->signal_monitor);
 	os_free(data->supp_freqs);
+	os_free(data->scan_freqs);
 	os_free(data);
 }
 
-- 
2.2.0.rc0.207.ga3a616c

